### PR TITLE
ITE drivers/sensor: add tachometer driver for it8xxx2_evb

### PR DESCRIPTION
**Pull request by RuibinChang**
Add tachometer driver for it8xxx2_evb.

Signed-off-by: Ruibin Chang <Ruibin.Chang@ite.com.tw>

### CODE REVIEW

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts
Comment (brockus-zephyr) [inline boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:18]: Does this have anything to do with adding the tach driver?  If not can you please make this a different commit?

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts
Comment (brockus-zephyr) [inline boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:42]: Does this have anything to do with adding the tach driver? If not can you please make this a different commit?

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (brockus-zephyr) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:114]: Can you create an IT8XXX2_PWM_xxxxx_MASK to remove the 0x0C magic number?

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (brockus-zephyr) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:122]: Can you create an IT8XXX2_PWM_xxxxx_MASK to remove the 0x03 magic number?

**Review by brockus-zephyr — CHANGES_REQUESTED** [review]

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts
Comment (RuibinChang) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:18]: Ok, I will create another commit for cleanup.

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts
Comment (RuibinChang) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts]: Ok, I will create another commit for cleanup.

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts
Comment (brockus-zephyr) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:18]: thanks

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts
Comment (brockus-zephyr) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts]: thanks

**Review by brockus-zephyr — APPROVED** [review]

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (MaureenHelm) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:132]: This should be `static`

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (MaureenHelm) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:153]: Please add braces even for single-line code blocks. See https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (MaureenHelm) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:188]: Can `data->capture` ever be zero?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig
Comment (RuibinChang) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig]: Ack. I put the conditional in the soc Kconfig.defconfig.series, because our Kconfig.defconfig sources from  Kconfig.defconfig.series.

Ok, I will create aonther PR for the rest.

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:149]: Done

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:185]: Yes,  1. when fan stop and tach data is valid, data->capture is zero, or
         2. tach data isn't valid, data->capture is zero.
 
But only data->capture > 0, we transform count to R.P.M by this formula, otherwise we set val->val1 = 0 directly.

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts
Comment (keith-zephyr) [inline boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:99]: If you added properties to the device tree to specify the TxDVS and TxCHSEL bits, then you could change the channel property to only track channel A/B.  This way it's not possible for the user to specify cross over the channel inputs.

The sample fetch routine could be simplified to
```
	if (tach_ch_valid(dev, tach_ch)) {
		/* If channel data of tachometer is valid, then save it */
		data->capture = ((*reg_fxtmrr) << 8) | (*reg_fxtlrr);
		/* Clear data valid */
		*reg_tswctlr |= config->dvs_bit;
```
Initialization would look like:
```
	if (tach_ch == TACH_CHANNEL_A) {
		/* Select TACHA output to tachometer */
		*reg_tswctlr &= ~config->chsel_bit;
		/* Clear tachometer data valid status */
		*reg_tswctlr |= config->dvs_bit;
```

The dvs_bit and chsel_bit properties would be set in the it8xxx2.dtsi file.

What do you think?

Comment (RuibinChang) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts]: Agree and we also needn't the tach_ch_to_pin_idx[] array.
Thanks a lot!

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts
Comment (MaureenHelm) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:18]: These changes still need to be moved to a different commit

File: boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig
Comment (MaureenHelm) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig]: Still need to remove `CONFIG_SENSOR=y` from `boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig`

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (MaureenHelm) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:108]: add braces here and around line 112

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:108]: Done

File: boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig
Comment (RuibinChang) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig]: We want to enable CONFIG_SENSOR on it8xxx2_evb board, if we remove CONFIG_SENSOR=y in boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig, then where shall I enable it?

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts
Comment (RuibinChang) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts]: Done

**Review by keith-zephyr — APPROVED** [review]

File: dts/bindings/tach/ite,it8xxx2-tach.yaml
Comment (mbolivar-nordic) [inline dts/bindings/tach/ite,it8xxx2-tach.yaml:14]: Please use dashes in property names, so `dvs-bit` here...

File: dts/bindings/tach/ite,it8xxx2-tach.yaml
Comment (mbolivar-nordic) [inline dts/bindings/tach/ite,it8xxx2-tach.yaml:19]: and `chsel-bit` here.

File: dts/bindings/tach/ite,it8xxx2-tach.yaml
Comment (mbolivar-nordic) [inline dts/bindings/tach/ite,it8xxx2-tach.yaml:34]: Putting the 0=A, 1=B information in the description rather than a comment would be better, so it shows up in https://docs.zephyrproject.org/latest/reference/devicetree/bindings.html

Not a blocking comment

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Minor DT changes requested

File: boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig
Comment (RuibinChang) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig]: Done

File: dts/bindings/tach/ite,it8xxx2-tach.yaml
Comment (RuibinChang) [reply dts/bindings/tach/ite,it8xxx2-tach.yaml]: Done

File: dts/bindings/tach/ite,it8xxx2-tach.yaml
Comment (RuibinChang) [reply dts/bindings/tach/ite,it8xxx2-tach.yaml]: Done

File: dts/bindings/tach/ite,it8xxx2-tach.yaml
Comment (RuibinChang) [reply dts/bindings/tach/ite,it8xxx2-tach.yaml]: Done

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (gmarull) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:64]: not related to this PR, but be aware that drivers should start moving to `pinctrl`, see https://github.com/zephyrproject-rtos/zephyr/issues/39740

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (gmarull) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:96]: please use `bool` here

non-blocking suggestion: `tach_ch_is_valid`

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (gmarull) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:133]: `SENSOR_CHAN_ALL` also needs to be supported in fetch

File: dts/riscv/it8xxx2.dtsi
Comment (gmarull) [inline dts/riscv/it8xxx2.dtsi:818]: consider using `BIT(n)` from `dt-util.h`

File: dts/riscv/it8xxx2.dtsi
Comment (gmarull) [inline dts/riscv/it8xxx2.dtsi:822]: this belongs to board-level DT, will likely need to be addressed when porting to pinctrl.

**Review by gmarull — CHANGES_REQUESTED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
My comments on this PR are addressed.

I am curious about one last thing, though:

```
			reg = <0x00f01820 1   /* F2TLRR */
			       0x00f01821 1   /* F2TMRR */
			       0x00f01848 1>; /* TSWCTLR */
```

Why `1` for the size? That's kind of weird value. Is the register block really 1 byte long?

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:64]: Ack

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

File: dts/riscv/it8xxx2.dtsi
Comment (RuibinChang) [reply dts/riscv/it8xxx2.dtsi]: Done

File: dts/riscv/it8xxx2.dtsi
Comment (RuibinChang) [reply dts/riscv/it8xxx2.dtsi:845]: Ack

**Comment by RuibinChang** [comment]
> My comments on this PR are addressed.
> 
> I am curious about one last thing, though:
> 
> ```
> 			reg = <0x00f01820 1   /* F2TLRR */
> 			       0x00f01821 1   /* F2TMRR */
> 			       0x00f01848 1>; /* TSWCTLR */
> ```
> 
> Why `1` for the size? That's kind of weird value. Is the register block really 1 byte long?

Our prescaler, PWM and tachometer module register are mess-up in 0xf01800~0xf018ff without any rules, so I list the register one by one for tachometer node.

**Comment by RuibinChang** [comment]
About this error: MAINTAINERS.yml: glob pattern 'drivers / * / * it8xxx2 *' in 'files' in area 'ITE Platforms' matches a directory, but has no trailing '/'
Why does the checker always check the original directory 'drivers / * / * it8xxx2 *' instead of added path 'drivers/ * / * / * it8xxx2 *' ?

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (gmarull) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:103]: `= false;`

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (gmarull) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:113]: ` = true;`

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (gmarull) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:108]: `true`

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (gmarull) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:133]: please validate chan here

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (gmarull) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:133]: you also need to support fetching `SENSOR_CHAN_ALL`, see https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/sensor.h#L508

File: include/dt-bindings/sensor/it8xxx2_tach.h
Comment (gmarull) [inline include/dt-bindings/sensor/it8xxx2_tach.h:11]: nit: better namespace these definitions (`IT8XXX2_...`), also add Doxygen docstrings:

```c
/**
 * @name Tachometer channels
 * @{
 */

/** Tachometer channel A */
#define TACH_CHANNEL_A		0
/** Tachometer channel B */
#define TACH_CHANNEL_B		1

/** @} */
```

File: soc/riscv/riscv-ite/common/chip_chipregs.h
Comment (gmarull) [inline soc/riscv/riscv-ite/common/chip_chipregs.h:48]: can the purpose/meaning be explained better here?

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

File: soc/riscv/riscv-ite/common/chip_chipregs.h
Comment (RuibinChang) [reply soc/riscv/riscv-ite/common/chip_chipregs.h]: Done

File: include/dt-bindings/sensor/it8xxx2_tach.h
Comment (RuibinChang) [reply include/dt-bindings/sensor/it8xxx2_tach.h]: Done

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

File: drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c
Comment (gmarull) [inline drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:133]: nit:

if wrong channel is provided, this function will still return 0, should return `-ENOTSUP`.

you can do:
```c
if ((chan != SENSOR_CHAN_RPM) && (chan != SENSOR_CHAN_ALL)) {
    return -ENOTSUP;
}
```

Comment (RuibinChang) [reply drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c]: Done

**Review by gmarull — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
For the binding

**Comment by RuibinChang** [comment]
Could we merge this PR? The PR https://github.com/zephyrproject-rtos/zephyr/pull/40274 needs the EC_FREQ common definition. Or need I do the changes for Compliance Checks / Check MAINTAINERS file?

**Comment by jackrosenthal** [comment]
> The PR #40274 needs the EC_FREQ common definition.

It seems like that PR depended on this one, but they were not uploaded together. Please revert that PR until this one lands.

**Comment by RuibinChang** [comment]
Ack

**Review by jackrosenthal — APPROVED** [review]

File: MAINTAINERS.yml
Comment (nashif) [inline MAINTAINERS.yml:1573]: what are you trying to match here? if `drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c`, then you should adapt the other wildcard to match the files explicitly, otherwise you will end up with a conflict.

Comment (RuibinChang) [reply MAINTAINERS.yml]: Yes, I want to match the drivers/.../tach_ite_it8xxx2.c path.
Done, thanks.

**Pull request merged**
